python/ray/serve/_private/http_proxy.py

```diff
@@ -565,6 +593,286 @@ async def send_request_to_replica_streaming(
+        response_proto = HealthzResponse(message=HEALTH_CHECK_SUCCESS_MESSAGE)
+        return ProxyResponse(
+            status_code=str(status_code),
+            response=response_proto.SerializeToString(),
+        )
+
+    def service_handler_factory(self, service_method: str, stream: bool) -> Callable:
+        async def unary_unary(
+            request_proto: Any, context: grpc._cython.cygrpc._ServicerContext
+        ) -> bytes:
+            """Entry point of the gRPC proxy unary request.
+
+            This method is called by the gRPC server when a unary request is received.
+            It wraps the request in a ProxyRequest object and calls proxy_request.
+            The return value is serialized user defined protobuf bytes.
+            """
+            proxy_request = gRPCProxyRequest(
+                request_proto=request_proto,
+                context=context,
+                service_method=service_method,
+                stream=False,
+            )
+            proxy_response = await self.proxy_request(proxy_request=proxy_request)
+            return proxy_response.response
+
+        async def unary_stream(
+            request_proto: Any, context: grpc._cython.cygrpc._ServicerContext
+        ) -> Generator[bytes, None, None]:
+            """Entry point of the gRPC proxy streaming request.
+
+            This method is called by the gRPC server when a streaming request is
+            received. It wraps the request in a ProxyRequest object and calls
+            proxy_request. The return value is a generator of serialized user defined
+            protobuf bytes.
+            """
+            proxy_request = gRPCProxyRequest(
+                request_proto=request_proto,
+                context=context,
+                service_method=service_method,
+                stream=True,
+            )
+            proxy_response = await self.proxy_request(proxy_request=proxy_request)
+            async for response in proxy_response.streaming_response:
+                yield response
+
+        if not stream:
+            return unary_unary
+        return unary_stream
+
+    async def send_request_to_replica_unary(
+        self,
+        handle: RayServeHandle,
+        proxy_request: ProxyRequest,
+    ) -> ProxyResponse:
+        return await self.send_request_to_replica_streaming(
+            request_id=proxy_request.request_id,
+            handle=handle,
+            proxy_request=proxy_request,
+        )
+
+    def setup_request_context_and_handle(
+        self,
+        app_name: str,
+        handle: RayServeHandle,
+        route_path: str,
+        proxy_request: ProxyRequest,
+    ) -> Tuple[RayServeHandle, str]:
+        """Setup request context and handle for the request.
+
+        Unpack gRPC request metadata and extract info to set up request context and
+        handle.
+        """
+        handle._set_request_protocol(RequestProtocol.GRPC)
```

Comment:
is it possible to refactor the code so that this is only called once when the handle is constructed instead of for every request?

Reply:
Yes, I have idea how to do that, let me try!

Reply:
The way this is done now is that the proxy router will be given the [protocol at initialization](https://github.com/ray-project/ray/pull/38609/files#diff-4dee6ac622ecd7c29896fc99c61c0bc1715281777d365d6cb4f1894172181ae5R31). Then when routes are updated, proxy router will just ensure the [protocol is set on the handle](https://github.com/ray-project/ray/pull/38609/files#diff-4dee6ac622ecd7c29896fc99c61c0bc1715281777d365d6cb4f1894172181ae5R69) Let me know if this approach looks good :)